distributed/shuffle/_disk.py

```diff
@@ -153,25 +153,31 @@ async def _process(self, id: str, shards: list[Any]) -> None:
         future then we should consider simplifying this considerably and
         dropping the write into communicate above.
         """
-        # Consider boosting total_size a bit here to account for duplication
-        with self.time("write"):
+        frames: Iterable[bytes | bytearray | memoryview]
+        if isinstance(shards[0], bytes):
+            # Manually serialized dataframes
+            frames = shards
+        else:
+            # Unserialized numpy arrays
+            with context_meter.meter("serialize", func=thread_time):
+                frames = list(concat(pickle_bytelist(shard) for shard in shards))
```

Comment:
I imagine this `list` call could have a non-trivial impact on performance since it forces the entire list of shards to be serialized and the serialized bytes to be retained before anything is written to disk, doesn't it?